ssl/quic/quic_stream_map.c

```diff
@@ -312,19 +345,251 @@ void ossl_quic_stream_map_update_state(QUIC_STREAM_MAP *qsm, QUIC_STREAM *s)
         stream_map_mark_inactive(qsm, s);
 }
 
+/*
+ * Stream Send Part State Management
+ * =================================
+ */
+
+int ossl_quic_stream_map_ensure_send_part_id(QUIC_STREAM_MAP *qsm,
+                                             QUIC_STREAM *qs)
+{
+    switch (qs->send_state) {
+    case QUIC_SSTREAM_STATE_NONE:
+        /* Stream without send part - caller error. */
+        return 0;
+
+    case QUIC_SSTREAM_STATE_READY:
+        /*
+         * We always allocate a stream ID upfront, so we don't need to do it
+         * here.
+         */
+        qs->send_state = QUIC_SSTREAM_STATE_SEND;
+        return 1;
+
+    default:
+        /* Nothing to do. */
+        return 1;
+    }
+}
+
+int ossl_quic_stream_map_notify_all_data_sent(QUIC_STREAM_MAP *qsm,
+                                              QUIC_STREAM *qs)
+{
+    switch (qs->send_state) {
+    default:
+        /* Wrong state - caller error. */
+    case QUIC_SSTREAM_STATE_NONE:
+        /* Stream without send part - caller error. */
+        return 0;
+
+    case QUIC_SSTREAM_STATE_SEND:
```

Comment:
Might be simpler just to do (same comment for next function):

```` C
if (qs->send_state != QUIC_SSTREAM_STATE_SEND)
    return 0;
````

Reply:
I could. I chose to do it like this as I think it makes the state-based handling more explicit/easier to understand.